# Assignment: Training and Evaluating a Small Transformer Language Model 

#### Course: Foundations of NLP

## 1. Setup

### 1.1 Clone the repository

In [5]:
!git clone https://github.com/karpathy/nanoGPT.git
%cd nanoGPT

f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT


fatal: destination path 'nanoGPT' already exists and is not an empty directory.


### 1.2 Install dependencies

In [6]:
!pip install torch numpy transformers datasets tiktoken wandb tqdm matplotlib

### 1.3 Prepare the dataset

In [7]:
%cd data/shakespeare_char
!python prepare.py
%cd ../..

f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT\data\shakespeare_char
length of dataset in characters: 1,115,394
all the unique characters: 
 !$&',-.3:;?ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz
vocab size: 65
train has 1,003,854 tokens
val has 111,540 tokens
f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT


## 2. Baseline Training Run

In [8]:
import torch

if torch.cuda.is_available():
    device_type = "cuda"
else:
    device_type = "cpu"

print("Using device:", device_type)

Using device: cuda


In [5]:
# Baseline configuration
out_dir = 'out-shakespeare-baseline'
eval_interval = 250
eval_iters = 200
log_interval = 10

always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda' # change to 'cuda' if you have a GPU
compile = False # set True only on Linux with GPU

Run the baseline:

In [9]:
!python train.py config/train_shakespeare_char_baseline.py

^C


### Generate text from the trained model

In [ ]:
!python sample.py --out_dir=out-shakespeare-baseline --device=cuda --num_samples=3 --max_new_tokens=500

Overriding: out_dir = out-shakespeare-baseline
Overriding config with \:


Traceback (most recent call last):
  File "f:\V.I.P\Uni\University-Documents-Repository\Foundation of NLP\Project\nanoGPT\sample.py", line 23, in <module>
    exec(open('configurator.py').read()) # overrides from command line or config file
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<string>", line 26, in <module>
FileNotFoundError: [Errno 2] No such file or directory: '\\'


## 3. Hyperparameter Experiments

Design  and  run  at  least  four  additional  experiments,  varying  one  hyperparameter  at  a  time relative to the baseline. You must cover at least three of the following categories:

| Category | Hyperparameter | Suggested values to try |
| - | - | - |
| Learning rate | learning_rate | 1e-4, 5e-4, 1e-3 (baseline), 5e-3, 1e-2 |
| Model depth | n_layer |2, 4, 6 (baseline), 8 |
| Model width | n_embd (n_head) | 128 (4 heads), 256 (4 heads), 384 (6 heads, baseline) |
| Context length | block_size | 64, 128, 256 (baseline) |
| Regularisation | dropout | 0.0, 0.1, 0.2 (baseline), 0.4 |
| Training duration | max_iters | 1000, 2500, 5000 (baseline), 10000 |

For each experiment, create a separate config file (e.g., config/train_shakespeare_char_lr5e4.py)  and  a  separate  out_dir  so  that  results  are  not overwritten.

Important: Keep all other hyperparameters fixed when varying one. Record the exact
configuration for every run.

In [ ]:
# Low Learningrate configuration
out_dir = 'out-shakespeare-low_lr'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-4 #CHANGE FROM BASELINE
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py config/train_shakespeare_char_low_lr.py

In [ ]:
# Less Layers configuration
out_dir = 'out-shakespeare-less_layers'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 256

n_layer = 2 #CHANGE FROM BASELINE
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py config/train_shakespeare_char_less_layers.py

In [ ]:
# Small Context configuration
out_dir = 'out-shakespeare-small_context'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 64 #CHANGE FROM BASELINE

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py config/train_shakespeare_char_small_context.py

In [ ]:
# Smaller Context configuration
out_dir = 'out-shakespeare-smaller_context'
eval_interval = 250
eval_iters = 200
log_interval = 10
always_save_checkpoint = False

wandb_log = False
wandb_project = 'nanoGPT-assignment'
wandb_run_name = 'baseline'

dataset = 'shakespeare_char'
gradient_accumulation_steps = 1
batch_size = 64
block_size = 128 #CHANGE FROM BASELINE

n_layer = 6
n_head = 6
n_embd = 384
dropout = 0.2

learning_rate = 1e-3
max_iters = 5000
lr_decay_iters = 5000
min_lr = 1e-4
beta2 = 0.99

warmup_iters = 100
weight_decay = 1e-1

device = 'cuda'   # change to 'cuda' if you have a GPU
compile = False   # set True only on Linux with GPU

In [ ]:
!python train.py config/train_shakespeare_char_smaller_context.py

## 4. Evaluation and Reporting

4.1 Quantitative evaluation

For each run, record:

• Final training loss

• Final validation loss

• Training time (wall-clock)

In [1]:
def readOutLogFile(folderPath, fileName):
    import os
    import numpy as np
    from string import printable

    logFileContent = ""

    filePath = os.path.join(folderPath, fileName)

    with open(filePath, 'r') as logFile:

        logFileContent = logFile.read()

    logData = {}
    logData["path"] = filePath
    logData["Experiment"] = fileName.replace("_", " ").replace(".log", "")
    logData["Iters"] = 5000
    
    itterationTimes = 0
    evaluationTimes = 0
    eval_interval = 250
    lastIteration = 0

    lines = logFileContent.split("\n")
    for line in lines:
        lineContext = ''.join(char for char in line if char in printable)
        lineContext = lineContext.replace(" ", "").replace("#CHANGEFROMBASELINE", "")
        if len(lineContext) > 1:
            if lineContext.startswith("learning_rate"):
                logData["LR"] = f"{float(lineContext.replace("learning_rate=", "")):.1e}"
            if lineContext.startswith("n_layer"):
                logData["Layers"] = int(lineContext.replace("n_layer=", ""))
            if lineContext.startswith("n_embd"):
                logData["Embd"] = int(lineContext.replace("n_embd=", ""))
            if lineContext.startswith("n_head"):
                logData["n_head"] = int(lineContext.replace("n_head=", ""))
            if lineContext.startswith("block_size"):
                logData["Block"] = int(lineContext.replace("block_size=", ""))
            if lineContext.startswith("dropout"):
                logData["Dropout"] = float(lineContext.replace("dropout=", ""))
            if lineContext.startswith("max_iters"):
                logData["Iters"] = int(lineContext.replace("max_iters=", ""))
            if lineContext.startswith("eval_interval"):
                eval_interval = int(lineContext.replace("eval_interval=", ""))
            if lineContext.startswith("step"):

                stepData = {}
                
                lineContextParts = lineContext.replace(":", ",").split(",")

                stepData["step"] = int(lineContextParts[0].replace("step", ""))
                stepData["trainloss"] = float(lineContextParts[1].replace("trainloss", ""))
                stepData["valloss"] = float(lineContextParts[2].replace("valloss", ""))

                if not ("steps" in logData):
                    logData["steps"] = []
                
                logData["steps"].append(stepData)
            if lineContext.startswith("iter"):

                iterData = {}
                
                lineContextParts = lineContext.replace(":", ",").split(",")

                iterData["iter"] = int(lineContextParts[0].replace("iter", ""))
                iterData["loss"] = float(lineContextParts[1].replace("loss", ""))
                iterData["time"] = float(lineContextParts[2].replace("time", "").replace("ms", ""))
                iterData["mfu"] = float(lineContextParts[3].replace("mfu", "").replace("%", ""))

                if not ("iter" in logData):
                    logData["iter"] = []

                if not (iterData["iter"] % eval_interval == 0):
                    itterationTimes += iterData["time"] * (iterData["iter"] - lastIteration)
                else:
                    evaluationTimes += iterData["time"]
                    if len(logData["iter"]) > 0 and iterData["iter"] - lastIteration - 1 >= 1:
                        itterationTimes += logData["iter"][-1]["time"] * (iterData["iter"] - lastIteration - 1)
                    
                logData["iter"].append(iterData)
                lastIteration = iterData["iter"]

    logData["Time (min)"] = round((itterationTimes + evaluationTimes) / 1000 / 60, 2)
    logData["Train Loss"] = logData["steps"][-1]["trainloss"]
    logData["Val Loss"] = logData["steps"][-1]["valloss"]

    return logData

In [2]:
logsFolderPath = "logs/"
logFileNames = ["baseline.log", "low_lr.log", "2_layers.log", "4_layers.log", "64_context.log", "128_context.log", "128_width.log", "256_width.log", "2500_iters.log", "1000_iters.log"]
tableColumns = ["Experiment", "LR", "Layers", "Embd", "Dropout", "Iters", "Train Loss", "Val Loss", "Time (min)"]

headderStr = "|"
seperatorStr = "|"
for column in tableColumns:
    headderStr += " " + "{:<11}".format(column) + " |"
    seperatorStr += "-" * 13 + "|"

print(headderStr)
print(seperatorStr)

for logFileName in logFileNames:
    logFileData = readOutLogFile(logsFolderPath, logFileName)
    
    rowStr = "|"

    for column in tableColumns:
        if column in logFileData:
            rowStr += " " + "{:<11}".format(str(logFileData[column])) + " |"
        else:
            rowStr += " |"

    print(rowStr)

| Experiment  | LR          | Layers      | Embd        | Dropout     | Iters       | Train Loss  | Val Loss    | Time (min)  |
|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
| baseline    | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 0.6232      | 1.6978      | 44.65       |
| low lr      | 1.0e-04     | 6           | 384         | 0.2         | 5000        | 1.1558      | 1.5021      | 44.63       |
| 2 layers    | 1.0e-03     | 2           | 384         | 0.2         | 5000        | 1.1023      | 1.4854      | 16.22       |
| 4 layers    | 1.0e-03     | 4           | 384         | 0.2         | 5000        | 0.8542      | 1.5319      | 29.97       |
| 64 context  | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 1.1576      | 1.4823      | 11.82       |
| 128 context | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 0.9449      | 1.50

#### Our Results:

| Experiment  | LR          | Layers      | Embd        | Dropout     | Iters       | Train Loss  | Val Loss    | Time (min)  |
|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|-------------|
| baseline    | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 0.6232      | 1.6978      | 44.65       |
| low lr      | 1.0e-04     | 6           | 384         | 0.2         | 5000        | 1.1558      | 1.5021      | 44.63       |
| 2 layers    | 1.0e-03     | 2           | 384         | 0.2         | 5000        | 1.1023      | 1.4854      | 16.22       |
| 4 layers    | 1.0e-03     | 4           | 384         | 0.2         | 5000        | 0.8542      | 1.5319      | 29.97       |
| 64 context  | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 1.1576      | 1.4823      | 11.82       |
| 128 context | 1.0e-03     | 6           | 384         | 0.2         | 5000        | 0.9449      | 1.5016      | 23.23       |
| 128 width   | 1.0e-03     | 6           | 128         | 0.2         | 5000        | 1.2719      | 1.5157      | 12.45       |
| 256 width   | 1.0e-03     | 6           | 256         | 0.2         | 5000        | 0.9703      | 1.4646      | 24.06       |
| 2500 iters  | 1.0e-03     | 6           | 384         | 0.2         | 2500        | 0.9565      | 1.4929      | 22.86       |
| 1000 iters  | 1.0e-03     | 6           | 384         | 0.2         | 1000        | 1.2706      | 1.5231      | 9.61        |

### 4.2 Loss curves

Create a **Python script** that:

• Parses the training logs for each run (nanoGPT prints step, train loss, val loss to stdout).

• Plots **training loss vs. step** for all runs in a single figure (one line per run, with a legend).

• Plots **validation loss vs. step** in a second figure.

Save these as .png or .pdf files. **Tip:** Redirect output to a log file during training:

In [ ]:
!python train.py config/train_shakespeare_char_baseline.py | tee logs/baseline.log

In [ ]:
# OUR PYTHON CODE

### 4.3 Qualitative evaluation

For **each** experiment, generate at least 3 text samples (500 tokens each) and include them in an
appendix. In the main body of your report, discuss:

• Which configurations produce the most coherent or "Shakespeare-like" text?

• At what point do you see clear signs of underfitting (e.g., gibberish output)?

• Do you observe any signs of overfitting (low train loss, high val loss, repetitive output)?

• Is there a noticeable difference between models with different context lengths?

In [ ]:
!python sample.py --out_dir=out-shakespeare-baseline \ --device=cpu --num_samples=3 --max_new

In [ ]:
!python sample.py --out_dir=out-shakespeare-low_lr \ --device=cpu --num_samples=3 --max_new

In [ ]:
!python sample.py --out_dir=out-shakespeare-less_layers \ --device=cpu --num_samples=3 --max_new

In [ ]:
!python sample.py --out_dir=out-shakespeare-small_context \ --device=cpu --num_samples=3 --max_new

In [ ]:
!python sample.py --out_dir=out-shakespeare-smallest_context \ --device=cpu --num_samples=3 --max_new

### 4.4 Analysis questions

Answer the following in your write-up (2–4 sentences each):

• **1.**  What  happens  when  the  learning  rate  is  too  high?  Too  low?  How  does  this  manifest  in  the
loss curve and the generated text?

• **2.**  What  is  the  relationship  between  model  size  (layers/embedding)  and  validation  loss?  Is
bigger always better at this scale and dataset size?

• **3.** What role does dropout play? Compare the train/val loss gap across dropout values.

• **4.** Based on your experiments, what configuration would you recommend for this dataset, and
why?

#### 5. Bonus (optional)

Choose one or both of the following:

• **A.** Train on a **custom dataset** of your choice (minimum 500KB of text). Describe the dataset,
any preprocessing you did, and compare sample quality with Shakespeare.

• **B.**  Replace  the  character-level  tokeniser  with  your  own  **Byte  Pair  Encoding  (BPE)**  tokeniser.
Train it on the Shakespeare corpus (or your custom dataset), integrate it into the nanoGPT data
pipeline,  and  compare  training  dynamics  and  generation  quality  against  the  character-level
baseline.